### PR TITLE
Fix a couple of shape inference issues

### PR DESCRIPTION
**Pull request by gramalingam**
### Description
Fixes a couple of issues revealed by [Issue 5212]( https://github.com/onnx/onnx/issues/5212).

(i) The data-propagation logic was not handling function-calls correctly. It was reusing the same map (for tracking values) in the caller and callee. We need to create a new map for the callee, and propagate values corresponding to the inputs (and outputs) to (and back from) the callee map.

(ii) Recently the proto definition was extended to specify default-values of attributes in FunctionProto. This extends the inference logic to account for it. (Note that the checker may require further extensions to handle this extension, that's to be done separately.)

(iii) Add test-cases to test the fixes.

### Motivation and Context
Fixes a couple of issues revealed by [Issue 5212]( https://github.com/onnx/onnx/issues/5212)

### CODE REVIEW

File: onnx/defs/generator/old.cc [hunk 1]
```diff
@@ -12,6 +12,116 @@ This operator produces a constant tensor. Exactly one of the provided attributes
 or value_* must be specified.
 )DOC";
 
+void ConstantInference(InferenceContext& ctx) {
```
Comment (gramalingam) [inline onnx/defs/generator/old.cc:15]: Note: No change to this logic in the file in this PR. It just moves the type-and-shape-inference function for the Constant op into a separately defined function. It was hard to set breakpoints inside this inference function and step through it in a debugger when it is included in a macro (as before).

File: onnx/shape_inference/implementation.cc [hunk 2]
```diff
@@ -345,6 +346,56 @@ class ShapeInferenceImplBase {
...
+        auto it = caller_map.find(actual);
+        if (it != caller_map.end()) {
+          callee_map[formal] = it->second;
+        }
+      }
+    }
+  }
+
+  // Update a DataValueMap for a calling function from the DataValueMap of the callee
+  void bindValuesOnReturn(
```
Comment (xadupre) [inline onnx/shape_inference/implementation.cc:369]: The code is very similar to the previous one, maybe that would be worth merging hem.

Comment (gramalingam) [reply onnx/shape_inference/implementation.cc:369]: Yes, they are similar. But hard to merge them. One focuses on inputs, the other on outputs. One copies from caller to callee, the other from callee to caller. But, in either case, we need to check caller-side input/output to see if it is empty. We will need to add so many extra-parameters (to merge them) that understanding the logic will be harder. I think it is better in its current form.

Comment (xadupre) [reply onnx/shape_inference/implementation.cc:369]: This is what I had in mind. I probably missed something.

```c++
  void bindValuesOnCallOrReturn(
      const DataValueMap& caller_map,
      const NodeProto& caller,
      bool io,
      DataValueMap& callee_map,
      const FunctionProto& callee) {
    int num_io = io 
      ? (std::min)(caller.input_size(), callee.input_size())
      : (std::min)(caller.output_size(), callee.output_size());
    for (int i = 0; i < num_io; ++i) {
      const std::string& actual = io ? caller.input(i) : caller.output(i);
      if (!actual.empty()) {
        const std::string& formal = io ? callee.input(i) : callee.output(i);
        auto it = caller_map.find(actual);
        if (it != caller_map.end()) {
          callee_map[formal] = it->second;
        }
      }
    }
  }

  // Initialize a DataValueMap for a called function from the DataValueMap of the caller
  void bindValuesOnCall(
      const DataValueMap& caller_map,
      const NodeProto& caller,
      DataValueMap& callee_map,
      const FunctionProto& callee) {
    bindValuesOnCallOrReturn(caller_map, caller, true, callee_map, callee);
  }

  // Update a DataValueMap for a calling function from the DataValueMap of the callee
  void bindValuesOnReturn(
      const DataValueMap& callee_map,
      const FunctionProto& callee,
      DataValueMap& caller_map,
      const NodeProto& caller) {
    bindValuesOnCallOrReturn(callee_map, caller, false, caller_map, callee);
  }
```

Comment (gramalingam) [reply onnx/shape_inference/implementation.cc:369]: In the innermost 3 lines, you will also need to change `caller_map` to `io ? caller_map : callee_map`, and similarly for `callee_map`. I prefer to keep the current form, as it is more understandable and is easier to verify is correct when we examine the code.

Comment (gramalingam) [reply onnx/shape_inference/implementation.cc:369]: And similarly we need to do the same for `actual` and `formal` in the innermost 3 lines.

**Review by xadupre — APPROVED** [review]

File: onnx/defs/generator/old.cc (hunk 1 above)
Comment (jcwchen) [reply onnx/defs/generator/old.cc:15]: Not relevant to this PR: might be worthy to make ConstantInference as a common function with options to handle different versions of Constant. Just like https://github.com/onnx/onnx/blob/a1e97a1ff257c7e3b1ecd62dc4872fa79059a275/onnx/defs/tensor/utils.cc#L150

**Review by jcwchen — APPROVED** [review]
LGTM. Thank you for fixing this! Also confirmed this PR can help to resolve segfault issue for maf_gpt2_tiny.onnx from https://github.com/onnx/onnx/issues/5212.

**Pull request merged**